src/any_agent/frameworks/any_agent.py

```diff
@@ -183,24 +181,29 @@ async def run_async(
         """
         trace = AgentTrace()
         trace_id: int
-        instrumentation_enabled = instrument and self._instrumentor is not None
 
         # This design is so that we only catch exceptions thrown by _run_async. All other exceptions will not be caught.
         try:
             with self._tracer.start_as_current_span(
                 f"invoke_agent [{self.config.name}]"
             ) as invoke_span:
-                if instrumentation_enabled:
-                    trace_id = invoke_span.get_span_context().trace_id
+                if self._wrapper:
                     async with self._lock:
-                        # We check the locked `_running_traces` inside `instrument`.
-                        # If there is more than 1 entry in `running_traces`, it means that the agent has
-                        # already being instrumented so we won't instrument it again.
-                        self._running_traces[trace_id] = AgentTrace()
-                        self._instrumentor.instrument(
-                            agent=self,  # type: ignore[arg-type]
+                        trace_id = invoke_span.get_span_context().trace_id
+                        self._wrapper.callback_context[trace_id] = Context(
+                            current_span=invoke_span,
+                            trace=AgentTrace(),
+                            tracer=self._tracer,
+                            shared={},
                         )
 
+                        if len(self._wrapper.callback_context) == 1:
```

Comment:
Could it be??? Just in case, I'm definitely not sure. Is there an initial context entry to consider?

```suggestion
                        if len(self._wrapper.callback_context) == 0:
```